File: inkuire-common/src/main/kotlin/org/virtuslab/inkuire/model/SerializableDocumentable.kt
package org.virtuslab.inkuire.model

data class SDRI(
    val packageName: String?,
    val className: String?,
    val callableName: String?,
    val original: String
) {
    override fun toString(): String = original
}

data class SDFunction(
    val dri: SDRI,
    val name: String,
    val isConstructor: Boolean,
    val parameters: List<SDParameter>,
    val areParametersDefault: List<Boolean>,
    val type: SBound,
    val generics: List<SDTypeParameter>,
    val receiver: SDParameter?
)

data class SDParameter(
    val dri: SDRI,
    val name: String?,
    val type: SBound
)

data class SDTypeParameter(
    val dri: SDRI,
    val name: String,
    val bounds: List<SBound>
)

data class AncestryGraph(
    val dri: SDRI,
    val type: SBound,
    val projections: List<SProjection>
)

File: commonScala/shared/src/main/scala/org/virtuslab/inkuire/model/SerializableDocumentable.scala
package org.virtuslab.inkuire.model

case class SDRI(
  packageName:  Option[String],
  className:    Option[String],
  callableName: Option[String],
  original:     String
) {
  override def toString: String = original
}

case class SDFunction(
  dri:                  SDRI,
  name:                 String,
  isConstructor:        Boolean,
  parameters:           List[SDParameter],
  areParametersDefault: List[Boolean],
  `type`:               SBound,
  generics:             List[SDTypeParameter],
  receiver:             Option[SDParameter],
  location:             String
)

case class SDParameter(
  dri:    SDRI,
  name:   Option[String],
  `type`: SBound
)

case class SDTypeParameter(
  variantTypeParameter: SVariance,
  bounds:               List[SBound]
)

case class AncestryGraph(
  dri:        SDRI,
  `type`:     SBound,
  superTypes: List[SProjection]
)
